vendor/github.com/mongodb/mongo-tools-common/db/read_preferences.go

```diff
@@ -24,38 +25,62 @@ const (
 		"connection to mongos may produce inconsistent duplicates or miss some documents."
 )
 
-func Primary() *readpref.ReadPref          { return readpref.Primary() }
-func PrimaryPreferred() *readpref.ReadPref { return readpref.PrimaryPreferred() }
-func Nearest() *readpref.ReadPref          { return readpref.Nearest() }
+// NewReadPreference takes a string (command line read preference argument) and a ConnString (from the command line
+// URI argument) and returns a ReadPref. If both are provided, preference is given to the command line argument. If
+// both are empty, a default read preference of primary will be returned.
+func NewReadPreference(cmdReadPref string, cs *connstring.ConnString) (*readpref.ReadPref, error) {
+	// default to primary if a read preference isn't specified anywhere
+	if cmdReadPref == "" && (cs == nil || cs.ReadPreference == "") {
+		return readpref.Primary(), nil
+	}
+
+	var doc readPrefDoc
+	var err error
+	if cmdReadPref != "" {
+		doc, err = readPrefDocFromString(cmdReadPref)
+	} else {
+		doc = readPrefDocFromConnString(cs)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	mode, err := readpref.ModeFromString(doc.Mode)
+	if err != nil {
+		return nil, err
+	}
+
+	tagSet := tag.NewTagSetFromMap(doc.Tags)
+	return readpref.New(mode, readpref.WithTagSets(tagSet))
+}
+
+func readPrefDocFromConnString(cs *connstring.ConnString) readPrefDoc {
+	doc := readPrefDoc{
+		Mode: cs.ReadPreference,
+	}
 
-func ParseReadPreference(rp string) (*readpref.ReadPref, error) {
-	var mode string
-	var tagSet tag.Set
-	if rp == "" {
-		return readpref.Nearest(), nil
+	if len(cs.ReadPreferenceTagSets) == 0 {
+		// no tag sets
+		return doc
 	}
+
+	// take the last tag set
+	doc.Tags = cs.ReadPreferenceTagSets[len(cs.ReadPreferenceTagSets) - 1]
```

Comment:
We shouldn't be discarding extra tag sets. It's totally fine for the user to specify more than one. Instead, update the `readPrefDoc` to have a slice of tag sets and use `NewTagSetsFromMaps` instead.

Reply:
Based on a conversation with Seth, we've decided to keep command line read preference parsing the way it is right now to avoid breaking changes and add support for multiple tag sets later. The new commit separates out the connection string read pref parsing to account for both this and maxStaleness.

Reply:
Sounds good. I think we might want to add a TODO comment describing what's going on and mentioning the ticket associated with command line rationalization somewhere in the docstring of `NewReadPreference` (so we don't forget).